torch/utils/data/datapipes/iter/combinatorics.py

```diff
@@ -113,21 +115,46 @@ def set_shuffle(self, shuffle=True):
         return self
 
     def __iter__(self) -> Iterator[T_co]:
+        self.reset()
         if not self._enabled:
             for x in self.datapipe:
                 yield x
         else:
-            buffer: List[T_co] = []
             for x in self.datapipe:
-                if len(buffer) == self.buffer_size:
-                    yield ShufflerIterDataPipe.buffer_replace(buffer, x)
+                if len(self.buffer) == self.buffer_size:
+                    yield ShufflerIterDataPipe.buffer_replace(self.buffer, x)
                 else:
-                    buffer.append(x)
-            random.shuffle(buffer)
-            while buffer:
-                yield buffer.pop()
+                    self.buffer.append(x)
+            random.shuffle(self.buffer)
+            while self.buffer:
+                yield self.buffer.pop()
 
     def __len__(self) -> int:
         if isinstance(self.datapipe, Sized):
             return len(self.datapipe)
         raise TypeError("{} instance doesn't have valid length".format(type(self).__name__))
+
+    def reset(self) -> None:
+        self.buffer = []
+
+    def __getstate__(self):
+        if IterDataPipe.getstate_hook is not None:
+            return IterDataPipe.getstate_hook(self)
+        state = (
+            self.datapipe,
+            self.buffer_size,
+            self._enabled
+        )
+        return state
+
+    def __setstate__(self, state):
+        (
+            self.datapipe,
+            self.buffer_size,
+            self._enabled
+        ) = state
+        self.reset()
+
+    def __del__(self):
+        if hasattr(self, "buffer"):
+            self.buffer.clear()
```

Comment:
I think `self.buffer` always exists.

Reply:
Turns out it doesn't

Reply:
TBH, I think this function is redundant. The `buffer` will be clean up automatically.

Reply:
I will have another look. I might just remove it. I just worry about the case that `torchvision` ran into.

Reply:
If that's the case, let's move `self.buffer` before assertion.